### PR TITLE
extensions: add http bandwidth limit filter

### LINKED ISSUE #13604: how to config bandwidth limit
requirement: such as limit bandwidth when client upload file to server 20kb/s
i find a filter "envoy.config.filter.fault.v2" https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/fault/v2/fault.proto#config-filter-fault-v2-faultratelimit-fixedlimit.
can i use this envoy filter to solve this question?
if it can, how config?

### PR DESCRIPTION
**Pull request by nitgoy**
Commit Message: adds bandwidth limit filter core functionality and UTs
Additional Description: the filter smoothens the flow of data in both direction up to the specified bandwidth limit.
Risk Level: Low, new filter
Testing: UTs added. Adding more UTs and integration tests.
Docs Changes: Added
Release Notes: Added
Fixes #13604

### CODE REVIEW

**Comment by nitgoy** [comment]
Actually found this warning with docs build command:
`generated/rst/api-v3/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto.rst: WARNING: document isn't included in any toctree`
Not sure what am I missing since the file is auto-generated.

**Comment by adisuissa** [comment]
> Actually found this warning with docs build command:
> `generated/rst/api-v3/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto.rst: WARNING: document isn't included in any toctree`
> Not sure what am I missing since the file is auto-generated.

Not an expert, but I think the idea is that the entire doc will be a tree (not a forest), and so everything can be reached from the main docs page.
Maybe something similar to [envoy/docs/root/api-v3/config/accesslog/accesslog.rst](https://github.com/envoyproxy/envoy/blob/main/docs/root/api-v3/config/accesslog/accesslog.rst) will work in this case.

**Comment by nitgoy** [comment]
> > Actually found this warning with docs build command:
> > `generated/rst/api-v3/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto.rst: WARNING: document isn't included in any toctree`
> > Not sure what am I missing since the file is auto-generated.
> 
> Not an expert, but I think the idea is that the entire doc will be a tree (not a forest), and so everything can be reached from the main docs page.
> Maybe something similar to [envoy/docs/root/api-v3/config/accesslog/accesslog.rst](https://github.com/envoyproxy/envoy/blob/main/docs/root/api-v3/config/accesslog/accesslog.rst) will work in this case.

I see. It didn't get picked up because [http.rst] (https://github.com/envoyproxy/envoy/blob/main/docs/root/api-v3/config/filter/http/http.rst) only looks in the v3 folder but not v3alpha folder. Changed that one but not sure if it's acceptable. Please review.

**Comment by nitgoy** [comment]
@adisuissa The docs error got fixed. Any insight into the format_pre failure or how to reproduce it locally?

**Comment by adisuissa** [comment]
> @adisuissa The docs error got fixed. Any insight into the format_pre failure or how to reproduce it locally?

According to the pipelines errors:
```
mixed tabs and spaces: docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst
```

To reproduce locally, you cant try running: `ci/run_envoy_docker.sh 'ci/do_ci.sh format_pre'`

File: docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+* Bandwidth limiting :ref:`architecture overview <arch_overview_bandwidth_limit>`
+* :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.bandwidth_limit.v3alpha.BandwidthLimit>`
+* This filter should be configured with the name *envoy.filters.http.bandwidth_limit*.
+
+The HTTP Bandwidth limit filter limits the size of data flow to the max bandwidth set in the limit_kbps
+when the request's route or virtual host has a per filter
+:ref:`bandwidth limit configuration <envoy_v3_api_msg_extensions.filters.http.bandwidth_limit.v3alpha.BandwidthLimit>`.
+
+If the bandwidth limit has been exhausted the filter stops further transfer until more bandwidth gets allocated
+according to the fill_interval(default is 50 milliseconds). If the connection buffer fills up with accumulated
```
[the remark this answers is not in the document]
Comment (nitgoy) [reply docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst:15]: Each refresh will be 50Bps, the limit will remain 1kBps.

File: source/common/common/token_bucket_impl.h [hunk 2]
```diff
@@ -17,8 +19,14 @@ class TokenBucketImpl : public TokenBucket {
    * @param time_source supplies the time source.
    * @param fill_rate supplies the number of tokens that will return to the bucket on each second.
    * The default is 1.
+   * @param mutex supplies the mutex object to be used to ensure thread-safety when the token bucket
+   * is shared. By default the class will be thread-unsafe.
    */
-  explicit TokenBucketImpl(uint64_t max_tokens, TimeSource& time_source, double fill_rate = 1);
+  explicit TokenBucketImpl(uint64_t max_tokens, TimeSource& time_source, double fill_rate = 1,
```
[the remark this answers is not in the document]
Comment (nitgoy) [reply source/common/common/token_bucket_impl.h:25]: These changes are not needed, seems like a bad rebase. Reverted.

File: docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+.. code-block:: yaml
+
+  name: envoy.filters.http.bandwidth_limit
+  typed_config:
+    "@type": type.googleapis.com/envoy.extensions.filters.http.bandwidth_limit.v3alpha.BandwidthLimit
+    stat_prefix: http_bandwidth_limiter
+    enable_mode: DecodeAndEncode
+    limit_kbps: 100
+    fill_interval: 40
```
Comment (yanavlasov) [inline docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst:34]: Is this still correct after you have changed the filed to Duration? There more examples below.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
LGTM module small cleanup in the examples. Thanks for this work.

/wait

File: docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst (hunk 3 above)
Comment (nitgoy) [reply docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst:34]: great catch. fixing.

**Comment by nitgoy** [comment]
Seeing only this failure in the coverage build, not related to my changes. Any ideas?

//test/integration:quic_http_integration_test               (3/6 cached) FAILED in 1 out of 6 in 165.5s

**Comment by nitgoy** [comment]
> LGTM module small cleanup in the examples. Thanks for this work.
> 
> /wait

Fixed. Thanks for the review.

Seeing only single failure in the coverage build, unrelated to my changes. Any ideas?

//test/integration:quic_http_integration_test (3/6 cached) FAILED in 1 out of 6 in 165.5s

File: docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst (hunk 3 above)
Comment (nitgoy) [reply docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst]: This is fixed.

File: docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst [hunk 4]
```diff
@@ -0,0 +1,100 @@
+.. _config_http_filters_bandwidth_limit:
+
+HTTP Bandwidth limit
+====================
+
+* Bandwidth limiting :ref:`architecture overview <arch_overview_bandwidth_limit>`
+* :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.bandwidth_limit.v3alpha.BandwidthLimit>`
+* This filter should be configured with the name *envoy.filters.http.bandwidth_limit*.
```
Comment (phlax) [inline docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst:8]: could you make this an inline literal please - eg ` ``envoy.filters.http.bandwidth_limit`` `

File: docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst (hunk 1 above)
Comment (phlax) [inline docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst:15]: is fill_interval literal - ie code - if so, can you use ` `` `

File: docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  typed_config:
+    "@type": type.googleapis.com/envoy.extensions.filters.http.bandwidth_limit.v3alpha.BandwidthLimit
+    stat_prefix: http_bandwidth_limiter
+
+
+The route specific configuration:
+
+.. code-block:: yaml
+
+  route_config:
```
Comment (phlax) [inline docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst:50]: it would be preferable to use a `literalinclude` with a full envoy config and only these lines showing - it just means that this config will be validated and tested

**Review by phlax — CHANGES_REQUESTED** [review]
thanks @nitgoy - ive added a few nits

**Comment by yanavlasov** [comment]
LGTM for source code. I think @phlax suggestion about moving example into separate yaml file so it can be checked in CI to avoid bitrot for examples is great.

/wait

File: docs/root/configuration/http/http_filters/_include/bandwidth-limit-filter.yaml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+            virtual_hosts:
+            - name: local_service
+              domains: ["*"]
+              routes:
+              - match: { prefix: "/path/with/bandwidth/limit" }
+                route: { cluster: service_protected_by_bandwidth_limit }
+                typed_per_filter_config:
+                  envoy.filters.http.bandwidth_limit:
+                    "@type": type.googleapis.com/envoy.extensions.filters.http.bandwidth_limit.v3alpha.BandwidthLimit
+                    enable_mode: DecodeAndEncode
```
Comment (yanavlasov) [inline docs/root/configuration/http/http_filters/_include/bandwidth-limit-filter.yaml:25]: I think it needs `stat_prefix`

**Review by yanavlasov — CHANGES_REQUESTED** [review]
Looks like test caught example config errors.

/wait

File: docs/root/configuration/http/http_filters/_include/bandwidth-limit-filter.yaml (hunk 6 above)
Comment (nitgoy) [reply docs/root/configuration/http/http_filters/_include/bandwidth-limit-filter.yaml:26]: yup

**Comment by phlax** [comment]
docs rendered here https://storage.googleapis.com/envoy-pr/16358/docs/index.html

File: docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst [hunk 7]
```diff
@@ -0,0 +1,61 @@
+.. _config_http_filters_bandwidth_limit:
+
+HTTP Bandwidth limit
+====================
+
+* Bandwidth limiting :ref:`architecture overview <arch_overview_bandwidth_limit>`
+* :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.bandwidth_limit.v3alpha.BandwidthLimit>`
+* This filter should be configured with the name ``envoy.filters.http.bandwidth_limit``.
+
+The HTTP Bandwidth limit filter limits the size of data flow to the max bandwidth set in the limit_kbps
```
Comment (phlax) [inline docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst:10]: nit: make `limit_kbps` literal

File: docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+when the request's route or virtual host has a per filter
+:ref:`bandwidth limit configuration <envoy_v3_api_msg_extensions.filters.http.bandwidth_limit.v3alpha.BandwidthLimit>`.
+
+If the bandwidth limit has been exhausted the filter stops further transfer until more bandwidth gets allocated
+according to the ``fill_interval`` (default is 50 milliseconds). If the connection buffer fills up with accumulated
+data then the source of data will be readDisabled as described in the :ref:`flow control doc<faq_flow_control>`.
```
Comment (phlax) [inline docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst:16]: clicking through to the flow control doc i dont see any further info on what `readDisabled` means

**Review by phlax — CHANGES_REQUESTED** [review]
@nitgoy thanks for this - a couple of nits below

one general point you might consider - as you have included the whole config file - perhaps you might want to use `:emphasize-lines:` - just a suggestion

File: docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst (hunk 8 above)
Comment (nitgoy) [reply docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst:16]: thanks. Actually, I intended to link to this doc: https://github.com/envoyproxy/envoy/blob/main/source/docs/flow_control.md. Fixed.

Comment (phlax) [reply docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst:16]: could you make `readDisabled` literal in that case  - also use `:repo:` directive for linking repo (if you havent done so already)

Comment (nitgoy) [reply docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst]: Fixed the link but didn't make 'readDisabled' literal since it is not code.

**Comment by nitgoy** [comment]
> @nitgoy thanks for this - a couple of nits below
> 
> one general point you might consider - as you have included the whole config file - perhaps you might want to use `:emphasize-lines:` - just a suggestion

Done. Which build's output contains link to the rendered docs?

File: docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+* :ref:`v3 API reference <envoy_v3_api_msg_extensions.filters.http.bandwidth_limit.v3alpha.BandwidthLimit>`
+* This filter should be configured with the name ``envoy.filters.http.bandwidth_limit``.
+
+The HTTP Bandwidth limit filter limits the size of data flow to the max bandwidth set in the ``limit_kbps``
+when the request's route or virtual host has a per filter
+:ref:`bandwidth limit configuration <envoy_v3_api_msg_extensions.filters.http.bandwidth_limit.v3alpha.BandwidthLimit>`.
+
+If the bandwidth limit has been exhausted the filter stops further transfer until more bandwidth gets allocated
+according to the ``fill_interval`` (default is 50 milliseconds). If the connection buffer fills up with accumulated
+data then the source of data will be readDisabled as described in the :repo:`flow control doc<source/docs/flow_control.md>`.
```
Comment (phlax) [inline docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst:16]: ``readDisabled``

**Comment by phlax** [comment]
> Done. Which build's output contains link to the rendered docs?

the link above will redirect to the most recently rendered docs

File: docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+  The token bucket is shared across all workers, thus the limits are applied per Envoy process.
+
+Example configuration
+---------------------
+
+Example filter configuration for a globally disabled bandwidth limiter but enabled for a specific route:
+
+.. literalinclude:: _include/bandwidth-limit-filter.yaml
+    :language: yaml
+	:lines: 11-53
```
Comment (phlax) [inline docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst:28]: i think this is the mixed tabs/spaces fail - https://dev.azure.com/cncf/envoy/_build/results?buildId=76487&view=logs&j=c5dd2866-6ab3-5f3c-3a44-4cef0ec909b5&t=a9eb66d6-8944-5769-b3f7-476949dadcb8&l=60

you probably want your editor to autofix-on-save the glint checks

Comment (nitgoy) [reply docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst:28]: thanks

File: docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst (hunk 9 above)
Comment (nitgoy) [reply docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst:16]: while `readDisable(bool)` is code, 'readDisabled' is not. Do you still want me to mark it as a literal?

Comment (phlax) [reply docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst:16]: hmm - not sure exactly - maybe rewording it a bit will help - reading this without prior knowledge might be confusing - perhaps something like - "...then the source of data will have `readDisable(True)` set as described..."

Comment (nitgoy) [reply docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst]: Done. Hope that makes it clearer.

**Review by phlax — APPROVED** [review]
lgtm from docs pov - thanks @nitgoy

**Review by yanavlasov — DISMISSED** [review]

File: api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+    // Filter enabled only for incoming traffic.
+    Decode = 1;
+
+    // Filter enabled only for outgoing traffic.
+    Encode = 2;
+
+    // Filter enabled for both incoming and outgoing traffic.
+    DecodeAndEncode = 3;
+  }
```
Comment (lizan) [inline api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto:39]: https://developers.google.com/protocol-buffers/docs/style#enums

> Use CamelCase (with an initial capital) for enum type names and CAPITALS_WITH_UNDERSCORES for value names:

File: docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+Example configuration
+---------------------
+
+Example filter configuration for a globally disabled bandwidth limiter but enabled for a specific route:
+
+.. literalinclude:: _include/bandwidth-limit-filter.yaml
+    :language: yaml
+    :lines: 11-53
+    :emphasize-lines: 9-29
```
Comment (phlax) [inline docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst:29]: nit: `:emphasize-lines: 9-25`

File: docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    :caption: :download:`bandwidth-limit-filter.yaml <_include/bandwidth-limit-filter.yaml>`
+
+Note that if this filter is configured as globally disabled and there are no virtual host or route level
+token buckets, no bandwidth limiting will be applied.
+
+Statistics
+----------
+
+The HTTP bandwidth limit filter outputs statistics in the *<stat_prefix>.http_bandwidth_limit.* namespace.
```
Comment (phlax) [inline docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst:38]: this is kinda a literal i think - either way please make it:

```
 ``<stat_prefix>.http_bandwidth_limit.``
```

**Review by phlax — CHANGES_REQUESTED** [review]
apologies for last minute nits @nitgoy - i would really appreciate if you could fix the literal - it looks wrong with italics 

see here:

https://storage.googleapis.com/envoy-pr/ce53ad6/docs/configuration/http/http_filters/bandwidth_limit_filter.html?highlight=bandwidth_limit#statistics

File: docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst (hunk 13 above)
Comment (nitgoy) [reply docs/root/configuration/http/http_filters/bandwidth_limit_filter.rst:38]: sure

File: api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto (hunk 11 above)
Comment (nitgoy) [reply api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto:39]: Ok. Fixing.

**Review by yanavlasov — APPROVED** [review]

File: api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto (hunk 11 above)
Comment (nitgoy) [reply api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto]: Done. would it be something that should be caught by the format checks?

**Review by yanavlasov — DISMISSED** [review]

File: source/extensions/filters/http/bandwidth_limit/bandwidth_limit.cc [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+
+void BandwidthLimiter::updateStatsOnEncodeFinish() {
+  if (encode_latency_) {
+    encode_latency_->complete();
+    encode_latency_.reset();
+    getConfig()->stats().encode_pending_.dec();
+  }
+}
+
+const FilterConfig* BandwidthLimiter::getConfig() const {
```
Comment (lizan) [inline source/extensions/filters/http/bandwidth_limit/bandwidth_limit.cc:181]: nit: if this cannot return a null, (from code above they assume so, use reference)

File: api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+    DISABLED = 0;
+
+    // Filter enabled only for incoming traffic.
+    DECODE = 1;
+
+    // Filter enabled only for outgoing traffic.
+    ENCODE = 2;
+
+    // Filter enabled for both incoming and outgoing traffic.
+    DECODE_AND_ENCODE = 3;
```
[the remark this answers is not in the document]
Comment (nitgoy) [reply api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto:38]: I chose it since Envoy can be used as ingress or egress proxy where request/response could mean different things. Open to changing it if you think this doesn't make sense.

File: source/extensions/filters/http/bandwidth_limit/bandwidth_limit.cc (hunk 14 above)
Comment (nitgoy) [reply source/extensions/filters/http/bandwidth_limit/bandwidth_limit.cc:181]: good idea

**Review by phlax — DISMISSED** [review]
lgtm for docs, thanks for cleanups @nitgoy

File: api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto (hunk 15 above)
[the remark this answers is not in the document]
Comment (nitgoy) [reply api/envoy/extensions/filters/http/bandwidth_limit/v3alpha/bandwidth_limit.proto]: Renamed on further thoughts and after seeing request/response usage in current APIs.

**Review by yanavlasov — APPROVED** [review]

**Review by lizan — APPROVED** [review]
